kicker/main.c

```diff
@@ -1,302 +1,403 @@
 #include <stdbool.h>
 
-#include <avr/wdt.h>
 #include <avr/interrupt.h>
 #include <avr/io.h>
+#include <avr/wdt.h>
 #include <stddef.h>
 #include <stdlib.h>
 #include <util/delay.h>
 
-#include "HAL_attiny167.h"
-
-#include "kicker_config.h"
 #include "kicker_commands.h"
+#include "kicker_config.h"
 #include "pins.h"
+#include "util.h"
 
 // kicker parameters
-#define MAX_KICK_STRENGTH 15.0f //2^4 - 1
+#define MAX_KICK_STRENGTH 255.0f
 #define MIN_EFFECTIVE_KICK_FET_EN_TIME 0.8f
 #define MAX_EFFECTIVE_KICK_FET_EN_TIME 10.0f
 
-#define KICK_TIME_SLOPE \
-    (MAX_EFFECTIVE_KICK_FET_EN_TIME - MIN_EFFECTIVE_KICK_FET_EN_TIME)
+#define KICK_TIME_SLOPE                                                        \
+  (MAX_EFFECTIVE_KICK_FET_EN_TIME - MIN_EFFECTIVE_KICK_FET_EN_TIME)
+
+#define KICK_COOLDOWN_MS 2000
+#define PRE_KICK_SAFETY_MARGIN_MS 5
+#define POST_KICK_SAFETY_MARGIN_MS 5
 
-// How much time to give for the LT to stop charging the caps
-#define STOP_CHARGING_SAFETY_MARGIN_MS 5
-// How much time to give the FET to stop current flow
-#define STOP_FLOW_SAFETY_MARGIN_MS 5
-// Time after the kick to allow the caps to charge back up somewhat
-#define CHARGE_TIME_MS 2000
+#define NO_COMMAND 0
 
 // number of steps of resolution we want per millisecond
 #define TIMER_PER_MS 4
 
+#define BALL_SENSE_MAX_SAMPLES 5
+
 // calculate our TIMING_CONSTANT (timer cmp val) from the desired resolution
-#define CLK_FREQ 8000000  // after removing default CLKDIV8 prescale
-#define TIMER_PRESCALE \
-   8  // set by TCCR0B |= _BV(CS01) which also starts the timer
+#define CLK_FREQ 8000000 // after removing default CLKDIV8 prescale
+#define TIMER_PRESCALE                                                         \
+  8 // set by TCCR0B |= _BV(CS01) which also starts the timer
 #define MS_PER_SECOND 1000
 
 #define MAX_TIMER_FREQ (CLK_FREQ / TIMER_PRESCALE)
 #define DESIRED_TIMER_FREQ (TIMER_PER_MS * MS_PER_SECOND)
 
 #define TIMING_CONSTANT ((MAX_TIMER_FREQ / DESIRED_TIMER_FREQ) - 1)
 
-// get different ball reading for X * 10 us before switching
-#define BALL_SENSE_MAX_SAMPLES 5
+// get different ball reading for 20 * 100 us = 2 ms before switching
+// Used to time kick and chip durations, -1 indicates inactive state
+volatile int32_t pre_kick_cooldown_ = -1;
+volatile int32_t timer_cnts_left_ = -1;
+volatile int32_t post_kick_cooldown_ = -1;
+volatile int32_t kick_wait_ = -1;
 
-// Corresponds to the values of the kick_type_is_kick
-#define IS_KICK true
-#define IS_CHIP false
+//Check for chip type
+volatile bool kick_type_is_chip_ = false;
+volatile bool kick_type_is_chip_command_ = false;
```

Comment:
I don't see these used, but the purpose is so that during a kick/chip, if the kick type is changed, we don't swap which FET we are opening or closing.

AKA
```
command immediate kick
open kick fet
command immediate chip
close chip fet
```
If `kick_type_is_chip` is locked on calls to `kick()`, and all commands are saved into `kick_type_is_chip_command` then this problem is eliminated.